tiatoolbox/models/segmentation/semantic_segmentor.py

```diff
@@ -0,0 +1,905 @@
+            # assume resolution idx to be in the same order as L
+            merged_resolution = ioconfig.highest_input_resolution
+            merged_locations = locations
+            # ! location is wrt highest resolution, hence still need conversion
+            if ioconfig.save_resolution is not None:
+                merged_resolution = ioconfig.save_resolution
+                output_shape = wsi_reader.slide_dimensions(**output_resolution)
+                merged_shape = wsi_reader.slide_dimensions(**merged_resolution)
+                fx = merged_shape[0] / output_shape[0]
+                merged_locations = np.ceil(locations * fx).astype(np.int64)
+            merged_shape = wsi_reader.slide_dimensions(**merged_resolution)
+            # 0 idx is to remove singleton wihout removing other axes singleton
+            to_merge_predictions = [v[idx][0] for v in predictions]
+            sub_save_path = f"{save_path}.raw.{idx}.npy"
+            self.merge_prediction(
+                merged_shape[::-1],  # XY to YX
+                to_merge_predictions,
+                merged_locations,
+                save_path=sub_save_path,
+                free_prediction=True,
+            )
+
+    @staticmethod
+    def merge_prediction(
+        canvas_shape: Union[Tuple[int], List[int], np.ndarray],
+        predictions: List[np.ndarray],
+        locations: Union[List, np.ndarray],
+        save_path: Union[str, pathlib.Path] = None,
+        free_prediction: bool = True,
+    ):
+        """Merge patch-level predictions to form a 2-dimensional prediction map.
+
+        Args:
+            canvas_shape (:class:`numpy.ndarray`): HW of the supposed assembled image.
+
+            predictions (list): List of nd.array, each item is a prediction of
+                a patch, assuming to be of shape HWC.
+
+            locations (list): List of nd.array, each item is the location of
+                the patch at the same index within `predictions`. The location
+                is in the to be assembled canvas and of the form
+                (top_left_x, top_left_y, bottom_right_x, bottom_right_x).
+
+            save_path (str): Location to save the assembled image.
+
+            free_prediction (bool): If this is `True`, `predictions` will
+                be modified in place and each patch will be replace with `None`
+                once processed. This is to save memory when assembling.
+
+        """
+        canvas_shape = np.array(canvas_shape)
+
+        sample_prediction = predictions[0]
+
+        num_output_ch = 0
+        add_singleton = False
+        canvas_count_shape_ = tuple(canvas_shape)
+        canvas_cum_shape_ = tuple(canvas_shape)
+        if len(sample_prediction.shape) == 3:
+            num_output_ch = sample_prediction.shape[-1]
+            canvas_cum_shape_ += (num_output_ch,)
+            add_singleton = num_output_ch == 1
+
+        cum_canvas = np.lib.format.open_memmap(
```

Comment:
memmap will take up quite a bit of memory. Can we also have an option to save the downsampled version on the fly?

Reply:
merge resolution defined in IOConfig determine which size the final output will be. https://github.com/TIA-Lab/tiatoolbox/blob/2825d5abaed301149bb594d92b848d4799e568f8/tiatoolbox/models/segmentation/semantic_segmentor.py#L78